Reflect recent API changes in event and player endpoints

maxim-01 opened the pull request:
- Updated ban-related fields in line with recent changes
- Narrowed down fields in certain types for better IntelliSense
- Loosened param types for methods returning API routes to accept string where applicable

maxim-01 committed:
Update to match recent API changes

- Updated ban-related fields in line with recent changes
- Narrowed down fields in certain types for better IntelliSense
- Loosened param types for methods returning API routes to accept string here applicable

ShawnCZek commented:
Thank you very much for a pull request with so many improvements! I will review it in the following days.

I let CI run to trigger the ESLint validation. Can you please take a look at it and make changes so it follows the setup rules of the repository? Also, I am not the biggest fan of the pull request title; do you think you can come up with something more descriptive?

maxim-01 committed:
Fix ESLint violations

ShawnCZek commented on web/v2/index.ts:
```diff
@@ -185,7 +185,7 @@ export const APIWebRoutes = {
    * @returns APIResponse<APIPlayer>
    */
-  player(id: bigint) {
+  player(id: bigint | number | string) {
```
`bigint` should already cover `number`, I believe.

Also, I am not really a fan of allowing string in all these requests that require an identifier.

maxim-01 replied:
Removed string where numeric identifiers are expected. As for bigint/number, [neither is compatible with the other](https://www.typescriptlang.org/play/#code/GYVwdgxgLglg9mABABwE4zFAQjA5hqACgA8AuRAIzwIEpyA3OGAEwCgBvVxbxVAUyghUSCAgDOcADZ8AdJLi4SNANysAvq1ahIsBCnSYAciAC2FPqhLkwp86jqJGLDlx79BwxKLATpchUqqGqxoBDj4mIQAjCohBlDGZhbRYCpAA)

ShawnCZek replied:
Good to know. Thank you for providing the example, too!

ShawnCZek commented on web/v2/ban.ts:
```diff
@@ -26,12 +26,18 @@ export interface APIPlayerBan {
+  adminID: null;
 }
+
+export type APIPlayerBans = APIPlayerBan[];
```
I believe it is unnecessary to export this as an array; we would have to do it for more structures to keep it consistent.

ShawnCZek commented on web/v2/player/index.ts:
```diff
@@ -191,9 +179,10 @@ export interface APIPlayer {
   /**
-   * The date and time the ban will expire (UTC) or `null` if not banned or ban is permanent.
+   * @deprecated
+   * This field is no longer available and will be returned as `null`.
```
It would have been useful to add the version when this got deprecated.

ShawnCZek commented on web/v2/player/index.ts:
```diff
@@ -227,3 +216,31 @@ export interface APIPlayer {
   vtcHistory: APIPlayerCompanyHistory[] | null;
 }
+
+/** Properties specific to the player resource when VTC information is hidden */
+interface _VTCHistoryHidden extends Pick<_Base, 'displayVTCHistory' | 'vtcHistory'> {
+  displayVTCHistory: false;
+  vtcHistory: null;
+}
+
+/** Properties specific to the player resource when VTC information is public */
+interface _VTCHistoryPublic extends Pick<_Base, 'displayVTCHistory' | 'vtcHistory'> {
+  displayVTCHistory: true;
+  vtcHistory: APIPlayerCompanyHistory[];
+}
+type _VTCHistory = _VTCHistoryHidden | _VTCHistoryPublic;
+
+/** Properties specific to the player resource when ban information is hidden */
+interface _BansHidden extends Pick<_Base, 'displayBans' | 'bansCount'> {
+  displayBans: false;
+  bansCount: null;
+}
+
+/** Properties specific to the player resource when ban information is public */
+interface _BansPublic extends Pick<_Base, 'displayBans' | 'bansCount'> {
+  displayBans: true;
+  bansCount: number;
+}
+type _BanData = _BansHidden | _BansPublic;
```
I get why the union is used, although, cannot we use the same approach as you used for the traffic API, i.e., without these "temporary" interfaces? I feel like they only make the code messier and could possibly leak to the build. Furthermore, the attributes lack documentation. Lastly, if we decide to keep them, it would have been nice to make the interface documentation consistent with the rest of the code base.

maxim-01 replied:
Simplified the initial 3-step types into 2-steps and named them more specifically. As for property documentation, since all the types inherit from `Pick<Base, ...>` the documentation also carries over (at least in VSCode, but I assume it isn't much different in other editors).

ShawnCZek replied:
The 2-step types are indeed better; thank you for the change. And the documentation indeed carries over. I did not notice that they are commented earlier in the file already. :)

ShawnCZek commented on web/v2/player/patreon.ts:
Same as the comment for `web/v2/player/index.ts`.

maxim-01 replied:
Named the "intermediate" types more specifically and removed exports from intermediate types to make them inaccessible from other translation units. Couldn't figure out an easy way to "simplify" the types whilst avoiding too much duplication, however did add a `#region` for better separation between types regarding hidden versus public Patreon info.

ShawnCZek replied:
The Patreon structure is indeed too hard to simplify, considering its complexity. I wonder whether we should follow the complexity within the types, as it makes maintenance a little bit more complicated.

I am not the biggest fan of the `#region` comments, as they are not officially supported. For instance, JetBrains products do not support them out of the box, so for me, they do not have any meaning.

maxim-01 replied:
Removed the region folds, instead reflected the "regions" in the type names

ShawnCZek reviewed (changes requested):
First of all, I am very sorry for the delay in getting to this pull request. As these repositories are not in my main scope, I have to find some extra time to properly maintain them.

The changes look really nice overall. Thank you a lot for this pull request; I can definitely see a huge improvement. I have left a few comments regarding my concerns and requests for little changes.

maxim-01 committed:
Make changes requested in #2

- removed exported `APIPlayerBans` type
- removed string from argument types when numeric identifiers are expected
- merged types that did not require separation

ShawnCZek commented on traffic/v2/index.ts:
```diff
@@ -3,7 +3,7 @@ import type { APITrafficServerGameName } from './server';
 export * from './traffic';
 
-export const APITrafficVersion = '2';
+export const APITrafficVersion = '2' as const;
```
This change should be done to the website API, too, or should be reverted here.

https://github.com/TruckersMP/API-Types/blob/1e77acc152a85e81c15c6e00966d7e5c140020bb/web/v2/index.ts#L9

ShawnCZek commented on web/v2/player/index.ts:
```diff
@@ -124,7 +107,12 @@ export interface APIPlayerCompanyHistory {
+
+// #region APIPlayer
+
+/** Base interface containing property annotations */
```
A little bit of nitpicking, though, can we stick to multiline comments to remain consistent, please?

The same goes for `web/v2/player/patreon.ts`.

ShawnCZek reviewed (commented):
Thank you for the quick changes! I left a few little remarks before we can merge this.

ShawnCZek commented on web/v2/player/index.ts:
```diff
@@ -124,7 +107,12 @@ export interface APIPlayerCompanyHistory {
+// #region APIPlayer
+
+/** Base interface containing property annotations */
+interface _Base {
```
I just noticed that this is still called `_Base`. Maybe `_APIPlayerBase` is a more suitable name?

maxim-01 replied:
Must have missed that one when renaming the other one - fixed.

maxim-01 committed:
Additional general style changes requested in #2

- generalised `as const` where applicable
- renamed the base interface for `APIPlayer` to align with `APIPlayerPatreon`

ShawnCZek reviewed (approved):
The changes look good to me. Thank you for the fast iteration! And thank you once again for this extensive pull request improving the API types. 🚀 

I am currently outside the town, so I will make a release later. Just so you know: we will also be moving to the `@truckersmp` organization (instead of `@truckersmp_official`) because we finally got the name from npm after months of waiting.

3ventic commented on web/v2/index.ts:
```diff
@@ -126,7 +126,7 @@ export const APIWebRoutes = {
    * @returns APIResponse<APICompanyRoles>
    */
-  companyRoles(companyId: string | number) {
+  companyRoles(companyId: number) {
```
This should be consistent with the others including companyId and accept a string as well.

maxim-01 committed:
Fixed missed type in #5c28f32

ShawnCZek committed:
Fix line endings for web/v2/ban.ts

ShawnCZek commented:
Thank you once again for this pull request. This change has been released as a part of v1.1.0.
